lib/Sema/CSSimplify.cpp

```diff
@@ -2920,7 +2920,13 @@ ConstraintSystem::matchExistentialTypes(Type type1, Type type2,
 
   // Handle existential metatypes.
   if (auto meta1 = type1->getAs<MetatypeType>()) {
-    if (auto meta2 = type2->getAs<ExistentialMetatypeType>()) {
+    ExistentialMetatypeType *meta2;
+    if (auto existential = type2->getAs<ExistentialType>()) {
+      meta2 = existential->getConstraintType()->getAs<ExistentialMetatypeType>();
```

Comment:
This seems superfluous: we already have an early return for `ExistentialMetatypeType` in `ExistentialType::get`.

Reply:
Yeah, some of these are redundant now (I wrote this code before I made `ExistentialType::get` return early for some of the types that aren't "supported" yet in my follow up PR). I'm holding off on doing major cleanups on `main`, including making all of the code paths _only_ operate on `ExistentialType`, in case I need to cherry-pick bug fixes to `release/5.6`.

Reply:
Thank you for letting me know!

> in case I need to cherry-pick bug fixes to `release/5.6`

FYI, the following is miscompiling on main. Looks like we're parsing `any` in the initializer as a decl ref to `any<Storage>(_ mask: SIMDMask<Storage>) -> Bool`.

```swift
protocol P {
  associatedtype A
}
do {
  let meta: (any P).Type = (any P).self
}
```

Reply:
Thanks! Here's the fix: https://github.com/apple/swift/pull/40949